Switch LogCopier from ScheduledExecutorService to HashedWheelTimer

kadaan opened the pull request:
@sahilTakiar @ibuenros Can you please review this fix for #729. Also, add anyone else on your side who has had experience with this.

sahilTakiar commented:
@chavdar might be interested, more details are in this issue: #729

kadaan commented:
@chavdar @sahilTakiar @ibuenros Can you review this?

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskScheduler.java:
```diff
@@ -0,0 +1,81 @@
+
+  abstract void startImpl(Optional<String> name);
+
+  public final void schedule(T task, long interval, TimeUnit timeUnit) {
```
Should this return a key? or the cancellable task itself?

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskScheduler.java:
```diff
@@ -0,0 +1,81 @@
+
+  public void cancel(CancellableScheduledTask<K> task) {
+    if (this.cancellableScheduledTaskMap.remove(task.getKey(), task)) {
+      task.cancel();
```
What happens if `task.cancel()` fails but the task has already been removed from the map?

kadaan replied:
I agree it is a little weird.  The passed in `T` is of type `ScheduledTask` and has a key.  Therefore the ScheduledTask can be found in the current task list by it's key.  The `CancelableScheduledTask` is only there because the method of cancelation is different for each type of Scheduler.  I'm open to any ideas.

ibuenros replied:
I am a bit confused, which task list are you referrring? `cancellableScheduledTaskMap` or a different list?

kadaan replied:
I'm referring to `cancellableScheduledTaskMap`.

ibuenros replied:
My concern is, say `task.cancel()` throws a runtime exception, which is caught by the caller. If the caller attempts to cancel the task once again, the task will no longer be in the map, and nothing will be cancelled. Or am I understanding this wrong?

kadaan replied:
How about `task.cancel()` can return a boolean and if it returns `true` we remove it from the collection and return `true`; otherwise, leave it in the collection and return `false`.

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskScheduler.java:
```diff
@@ -0,0 +1,81 @@
+      try {
+        this.closeImpl();
+      } finally {
+        isStarted = false;
```
I think `isStarted` should be set to false before calling `closeImpl()`. The call might be slow, but we should block new tasks as soon as the close starts.

kadaan replied:
How about adding a isClosing and setting that?  I don't want it to be able to be started while closing.

ibuenros replied:
Is there a reason to support restarting? IMO once it's closed, it should never be restarted, so you could just have a three state version instead: `unstarted`, `started`, `closed`.

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskSchedulerFactory.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Map;
+
+public class TaskSchedulerFactory {
+  private static final Map<String, Class> SCHEDULER_MAP = buildSchedulerMap();
```
We try to avoid static maps, as they may generate memory leaks. An alternative is to use a Cache.

kadaan replied:
You are referring to the Guava cache, which can be size bounded, right?  

In this case, how can it leak?

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskSchedulerFactory.java:
```diff
@@ -0,0 +1,54 @@
+        }
+      }
+    }
+    return get();
```
Can this be handled in this method instead of recursing?

kadaan replied:
Sure.

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskSchedulerFactory.java:
```diff
@@ -0,0 +1,54 @@
+
+import java.util.Map;
+
+public class TaskSchedulerFactory {
```
Can you explain what the rationale for this class is? It seems to be caching something, but I don't understand what.

kadaan replied:
This is so the TaskSchedulers can be loaded by name.  I want to be able to allow the TaskScheduler to be configured in the gobblin config so that we can switch between them before we are certain that the new Scheduler is performing correctly.

ibuenros replied:
Then I would rather use an enum instead of a map (also solves the issue with the static map).

kadaan replied:
I can do that.

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/ScheduledTask.java:
```diff
@@ -0,0 +1,19 @@
+public interface ScheduledTask<K> {
+    K getKey();
+
+    void run();
```
Should we allow exceptions here?

kadaan replied:
I want to ensure that now exceptions can leak out other than RuntimeExceptions because depending on the implementation of the TaskScheduler, it cause break things.  For example, `ScheduledTaskExecutor` indicates: "...if any execution of the task encounters an exception, subsequent executions are suppressed..."

ibuenros commented:
Can you add javadoc and copyright block to all classes?

kadaan commented:
@ibuenros Updated based on your comments.

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskSchedulerFactory.java:
```diff
@@ -0,0 +1,113 @@
+    @Getter
+    private final Class<? extends TaskScheduler> taskSchedulerClass;
+
+    public static Optional<TaskSchedulerType> parse(Optional<String> name) {
```
What is the point of returning an optional here? Couldn't you just return the default? (that seems to be the only way you are using `parse`)

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskSchedulerFactory.java:
```diff
@@ -0,0 +1,113 @@
+   * @param <T> the type of the {@link ScheduledTask}
+   * @return an instance of {@link TaskScheduler}
+   */
+  public static <K, T extends ScheduledTask<K>> TaskScheduler<K, T> get(String type) {
```
This method's signature is too close to `get(Optional<String>)`, it will probably cause confusion in the future.

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskSchedulerFactory.java:
```diff
@@ -0,0 +1,113 @@
+    return getImpl(Optional.fromNullable(type), name);
+  }
+
+  private static <K, T extends ScheduledTask<K>> TaskScheduler<K, T> getImpl(Optional<String> type, Optional<String> name) {
```
Should this signature be in fact `getImpl(String, Optional<String>)`? Given that type is not actually optional? The methods with no type can get the default type themselves.

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskScheduler.java:
```diff
@@ -0,0 +1,134 @@
+  public final void schedule(final T task, final long period, final TimeUnit unit) throws IOException {
+    Preconditions.checkArgument(isStarted, "TaskScheduler is not started");
+    try {
+      this.cancellableTaskMap.get(task.getKey(),
```
If a task with that key is already present, this will silently not schedule the new task. I think it should throw an exception instead.

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskScheduler.java:
```diff
@@ -0,0 +1,134 @@
+ * Base class for a task scheduler that can run {@link ScheduledTask}s periodically. Subclasses can implement
+ * {@link #startImpl(Optional)} and {@link #closeImpl()}.
+ */
+public abstract class TaskScheduler<K, T extends ScheduledTask<K>> implements Closeable {
```
Is thread safety a concern? If so, I believe most methods in this class are not thread safe. If not, please specify that in the javadoc.

kadaan replied:
`start` and `close` are thread-safe because they are synchronized.  `schedule` is thread-safe because the underlying `get` call is thread-safe.

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskScheduler.java:
```diff
@@ -0,0 +1,134 @@
+ * {@link #startImpl(Optional)} and {@link #closeImpl()}.
+ */
+public abstract class TaskScheduler<K, T extends ScheduledTask<K>> implements Closeable {
+  private final Cache<K, CancellableTask<K, T>> cancellableTaskMap = CacheBuilder.newBuilder().build();
```
Is there a reason to use a `Cache` instead of a `ConcurrentMap`?

kadaan replied:
Because cache has a thread-safe computeIfAbsent method that is Java7 compatible: `get(K, Callable<? extends V>)`

ibuenros commented:
Please add javadoc to all classes.

kadaan commented:
@ibuenros Updated based on your comments.  

1. Removed a bunch of overrides for `TaskSchedulerFactory`, and added one that takes the enum rather than the string.  
1. Made the `TaskSchedulerType` enum public.  
1. Indicated that `TaskScheduler` implementations should be thread-safe. 
1. Added javadocs for all non-test classes and method (even private and package visibility). 
1. Fail the `schedule` call if the task cannot be scheduler

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/concurrent/TaskScheduler.java:
```diff
@@ -0,0 +1,165 @@
+   * @param task the {@link ScheduledTask} to cancel
+   * @return true if the {@link ScheduledTask} was canceled; otherwise, false
+   */
+  public boolean cancel(T task) {
```
This call is not thread-safe. While the task is canceling, a different thread might cancel the same task.

kadaan replied:
I don't think it matters for the following reasons:

1. The underlying `cancellableTaskMap` is thread safe
1. The `ScheduledTask.cancel` is safe to call multiple times
1. Until the `this.cancellableTaskMap.invalidate(task.getKey())` call, you cannot create another task with the same key.

The only improvement I can see is to add a reference check against the `T task` and the `cancellableTask.getScheduledTask()` to ensure that the passed in task is the same one that will be cancelled.  If not, just return false (because the passed in task has not been scheduled).

ibuenros replied:
Why is it necessary safe to call `cancel()` multiple times? (The interface does not specify that, and even if it did, it is very easy for users to ignore that requirement).
The check that the task to cancel is the expected one is another good thing to add.

kadaan replied:
Well, all implementations of the cancel method are currently safe.  For instance the `HashedWheelTimer` `Timeout` states: "If the task has been executed or cancelled already, it will return with no side effect."  Additionally, the `Future` states the `cancel` method returns: "false if the task could not be cancelled, typically because it has already completed normally".  I don't think it is unreasonable for interfaces to require that the implementations are thread-safe.  This is done in many other places.  My preference is to reduce high level locking as much as possible.

ibuenros replied:
Ok, at least specify in the interface that multiple calls to cancel must be safe.

kadaan replied:
FYI, the implementation of `CancellableTask` for `ScheduledExecutorServiceTaskScheduler` does not synchronize the `cancel` method because it isn't needed.  The implementation of `CancellableTask` for `HashedWheelTimerTaskScheduler` does synchronize the `cancel` method to ensure that the future isn't being set by the `runOneIteration` method and unset by `cancel` at the same time.

ibuenros commented:
Just one more comment.

kadaan commented:
@ibuenros Updated with comment

ibuenros commented:
LGTM

kadaan commented:
@ibuenros Thanks!
